tests/functional/ovndbcluster_controller_test.go

```diff
@@ -27,11 +28,121 @@ import (
+					Eventually(func(g Gomega) {
+						c := GetOVNDBCluster(clusterName)
+						*c.Spec.Replicas = 1
+						g.Expect(k8sClient.Update(ctx, c)).Should(Succeed())
+					}).Should(Succeed())
+
+					// Check if dnsdata CR is down to 1
+					Eventually(func() int {
+						targeted_dns_ocurrences := 0
+						listDns := GetDNSDataList(statefulSetName)
+						// This calls CreateOVNDBClusters at the BeforeEach, which will
+						// create NB and SB cluster, both belonging at the same namespace
+						// GetDNSDataList will return ALL occurrences of DNSData (will include
+						// the NB and the SB occurences). Since both clusters were created with 3
+						// replicas and only the targeted one has been scaled down to 1 if we don't
+						// filter by name the result would be 4 (3+1)
+						for _, d := range listDns.Items {
+							if strings.Contains(d.Name, DnsName) {
+								targeted_dns_ocurrences++
+							}
+						}
+						return targeted_dns_ocurrences
+					}).Should(BeNumerically("==", 1))
+				},
+				Entry("DNS entry NB", "sb"),
+				Entry("DNS entry SB", "nb"),
+			)
+			It("Should update DNSData IP if pod IP changes", func() {
+				var clusterName types.NamespacedName
+				var OVNSBDBClusterName types.NamespacedName
+				var cluster *ovnv1.OVNDBCluster
+				var dbs []types.NamespacedName
+				_ = th.CreateNetworkAttachmentDefinition(types.NamespacedName{Namespace: namespace, Name: "internalapi"})
+				dbs = CreateOVNDBClusters(namespace, map[string][]string{namespace + "/internalapi": {"10.0.0.1"}}, 3)
+				OVNSBDBClusterName = types.NamespacedName{Name: dbs[1].Name, Namespace: dbs[1].Namespace}
+				cluster = GetOVNDBCluster(OVNSBDBClusterName)
+				clusterName = OVNSBDBClusterName
+				// DnsEntry will have two entries, the ovsdbserver-sb.namespace.svc
+				// and the ovsdbserver-sb-0.namespace.svc, both will have same IP
+				// for this test ovsdbserver-sb-0 will be used.
+				Eventually(func(g Gomega) {
+					dnsEntry := GetDNSData(types.NamespacedName{Name: "dns-ovsdbserver-sb-0", Namespace: cluster.Namespace})
+					g.Expect(dnsEntry).To(Equal("10.0.0.1"))
+				})
+
+				// Modify pod IP section
+				pod := GetPod(types.NamespacedName{Name: "ovsdbserver-sb-0", Namespace: cluster.Namespace})
+				// Create new pod NAD and add it to POD
+				var netStatus []networkv1.NetworkStatus
+				var IPString []string
+				IPString = append(IPString, "10.0.0.10")
+				netStatus = append(
+					netStatus,
+					networkv1.NetworkStatus{
+						Name: cluster.Namespace + "/internalip",
+						IPs:  IPString,
+					},
+				)
+				netStatusAnnotation, err := json.Marshal(netStatus)
+				Expect(err).NotTo(HaveOccurred())
+				pod.Annotations[networkv1.NetworkStatusAnnot] = string(netStatusAnnotation)
+				UpdatePod(pod)
+				// End modify pod IP section
+
+				// Call reconcile loop
+				Eventually(func(g Gomega) {
+					c := GetOVNDBCluster(clusterName)
+					// Change something just to call reconcile loop
+					*&c.Spec.ElectionTimer = 11000
```

Comment:
I don't think we should watch the operator directly, as on a real environment the NAD IP will only change on destroy/creation of the pod. This will only happens during a reschedule of the pod or due to pod being on a bad state (Let me know if there's other scenarios that I'm not taking into consideration). Both this scenarios are something that can't be replicated on the tests. So I think that workaround is a simple way to trigger the reconcile.

Reply:
OK. I assumed that pod IP may never change, but it looks like it's not really guaranteed: https://github.com/kubernetes/kubernetes/issues/108281#issuecomment-1048243795

Then let's keep your workaround to simulate the pod ip change mid-flight. It's not a normal situation, but worth being handled.